design/cluster-update.md

```diff
@@ -0,0 +1,150 @@
+For example, if `monCount` were to be updated, the operator would need to add or remove a single monitor at a time, ensuring that quorum is restored before moving onto the next monitor.
+Updates to the storage spec for the cluster require even more careful consideration and management by the operator, which will be discussed in this section.
+
+It is important to remember that a single OSD pod can contain multiple OSD processes and that the operator itself does not have detailed knowledge of the storage resources of each node.
+More specifically, the devices that can be used for storage (e.g., match `deviceFilter`) is not known until the OSD pod has been started on a given node.
+
+As mentioned previously, it is recommended to make storage changes to the cluster one OSD at a time.
+Therefore, the operator and the OSD pods will need to coordinate their efforts in order to adhere to this guidance.
+When a cluster update event is received by the operator, it will work on a node by node basis, ensuring all storage updates are completed by the OSD pod for that node before moving to the next.
+
+When an OSD pod starts up and has completed its device discovery, it will need to perform a diff of the desired storage against the actual storage that is currently included in the cluster.
+This diff will determine the set of OSD instances that need to be removed or added within the pod.
+Fortunately, the OSD pod start up is already idempotent and already handles new storage additions, so the remaining work will be the following:
+* Safely removing existing OSDs from the cluster
+* Waiting for data migration to complete and all placement groups to become clean
+* Signaling to the operator that the pod has completed its storage updates
+
+We should consider an implementation that allows the OSD pod to refresh it's set of OSDs without restarting the entire pod, but since the replication controller's pod template spec needs to be updated by the operator in order to convey this information to the pod, we may need to live with restarting the pod either way.
+Remember that this will be done one node at a time to mitigate impact to cluster health.
+
+Also, other types of update operations to the cluster (e.g., software upgrade) should be blocked while a cluster update is ongoing.
+
+**TODO:** how does OSD pod signal that it is done with OSD changes?  Perhaps setting a value in a config map?
+
+### Storage Update Process
+This section covers the general sequence for updating storage resources and outlines important considerations for cluster health.
+Before any changes begin, we will temporarily disable scrubbing of placement groups (the process of verifying data integrity of stored objects) to maximize cluster resources that can go to both client I/O and recovery I/O for data migration:
+```console
+ceph osd set noscrub
+ceph osd set nodeep-scrub
```

Comment:
I think `ceph osd set noout`  should also be run/set (see http://docs.ceph.com/docs/giant/rados/troubleshooting/troubleshooting-osd/#stopping-w-out-rebalancing).

Reply:
i'm not sure if we should do the `set noout`, because for removal we do want OSDs to be marked as out so their data gets rebalanced.  For addition, I'm not sure if `noout` helps with anything specifically.  Am I not thinking of a certain case @galexrt?